[change:ux] Added command loading overlay #714

Aryamanz29 opened the pull request:
![Screenshot from 2022-12-07 16-01-53](https://user-images.githubusercontent.com/56113566/206155895-273504d6-5e70-4e00-93ce-bfd42cfa7a4e.png)

Closes #714

nemesisdesign reviewed (changes requested):
This works pretty well, except when sending the first command of a device which doesn't have any other recent commands listed in the recent commands tab, in that case is not shown, can you please look into that as well to maintain consistency?

Aryamanz29 commented on openwisp_controller/connection/static/connection/js/commands.js:
```diff
@@ -451,7 +451,7 @@ function initCommandOverlay($) {
                     </div>
                     <div class="form-row field-output">
-                        <div><label>Output:</label><div class="readonly">${response.output}</div></div>
+                        <div><label>Output:</label><div class="readonly">${response.status === 'in-progress'? `<div class="loader"></div>` : response.output}</div></div>
```
Sure :+1:

nemesisdesign reviewed (commented):
Looks good, will test tomorrow.

nemesisdesign reviewed (changes requested):
That's not what I meant, I meant to add it here:

![Screenshot from 2022-12-12 23-15-49](https://user-images.githubusercontent.com/841044/207209606-74a23308-08f2-4ca3-846d-8b49764ab259.png)

nemesisdesign reviewed (changes requested):
Let me explain with a recording so it becomes clear.

![Peek 2022-12-14 20-32](https://user-images.githubusercontent.com/841044/207738092-8faf8dbb-c1c6-462e-bf54-4957260ee265.gif)

Here's what happens:

- user submits command, there's no recent commands tab yet, so the form submits and the page reloads
- after the page reloads the loading overlay is not shown

I see that you are showing the command output by replacing the entire HTML content, please revert that.

I think the loading overlay should be shown once the page refreshes, instead of the empty status content which is being shown now.

Is what I am asking clearer? If not, let's discuss.

Aryamanz29 commented:
> Let me explain with a recording so it becomes clear.
> Is what I am asking clearer? If not, let's discuss.

@nemesisdesign, It appears slightly different in your environment; I also tested the same command, and the command was successfully executed without showing me an empty output container, as shown below:

https://user-images.githubusercontent.com/56113566/208458118-c18a0dc4-9922-4cf3-acb1-d8577e60f63d.mp4

### `Updates`:

- Okay, I found out that when we reload the page in the `"in-process"` state, I was able to reproduce what you've mentioned above.

https://user-images.githubusercontent.com/56113566/208459964-f1002d5b-9ae1-4cf9-8261-e69e9f6298aa.mp4

### `Possible solution`:

- When I looked into the code, I discovered that command results are[ received via websocket](https://github.com/openwisp/openwisp-controller/blob/2b978723bf793aa40c88bd2d8f8e9bc8cd26c1bf/openwisp_controller/connection/channels/consumers.py#L10-L13) and we do an ajax request before building `firstCommandElement`, and I've added the loading overlay to that element only, but if someone reloads that page in between the `"in-process"` command state, we'll get an empty output container because we're now getting data directly through DB as [defined in admin.py](https://github.com/openwisp/openwisp-controller/blob/2b978723bf793aa40c88bd2d8f8e9bc8cd26c1bf/openwisp_controller/connection/admin.py#L92-L134), So the possible fix to this problem is that we can customise our admin output field to add a loading overlay if the command objects have `status == "in-process"`.

- Added in [`f4ad9a0` (#727)](https://github.com/openwisp/openwisp-controller/pull/727/commits/f4ad9a07f869e048f97b4067da82b74c985f6c22)

https://user-images.githubusercontent.com/56113566/208464240-22bab449-bac1-406d-a597-5218054afda6.mp4
